src/test/java/com/thomashan/collection/immutable/ImmutableListImplTest.java

```diff
@@ -0,0 +1,253 @@
+
+        assertThrows(UnsupportedOperationException.class, () -> of(object).retainAll(Collections.singleton(object)));
+    }
+
+    @Test
+    public void testClear_ThrowsUnsupportedOperationException() {
+        assertThrows(UnsupportedOperationException.class, () -> of().clear());
+    }
+
+    @Test
+    public void testGet_GivenObjectIsInTheList_ReturnsObject() {
+        Object object = new Object();
+
+        assertEquals(object, of(object).get(0));
+    }
+
+    @Test
+    public void testGet_GivenObjectIsNotInTheList_ThrowsIndexOutOfBoundsException() {
+        assertThrows(IndexOutOfBoundsException.class, () -> of().get(0));
+    }
+
+    @Test
+    public void testAddWithIndex_ThrowsUnsupportedOperationException() {
+        assertThrows(UnsupportedOperationException.class, () -> of().add(0, new Object()));
+    }
+
+    @Test
+    public void testRemoveWithIndex_ThrowsUnsupportedOperationException() {
+        assertThrows(UnsupportedOperationException.class, () -> of(new Object()).remove(0));
+    }
+
+    @Test
+    public void testIndexOf_GivenObjectIsInTheList_ReturnsFirstIndex() {
+        Object object = new Object();
+
+        assertEquals(0, of(object, object).indexOf(object));
+    }
+
+    @Test
+    public void testIndexOf_GivenObjectIsNotInTheList_ReturnsNegativeOne() {
+        assertEquals(-1, of().indexOf(new Object()));
+    }
+
+    @Test
+    public void testLastIndexOf_GivenObjectIsInTheList_ReturnsLastIndex() {
+        Object object = new Object();
+
+        assertEquals(1, of(object, object).lastIndexOf(object));
+    }
+
+    @Test
+    public void testLastIndexOf_GivenObjectIsNotInTheList_ReturnsNegativeOne() {
+        assertEquals(-1, of().lastIndexOf(new Object()));
+    }
+
+    @Test
+    public void testListIterator_ReturnsListIterator() {
+        assertTrue(of().listIterator() instanceof ListIterator);
+    }
+
+    @Test
+    public void testListIteratorWithIndex_ReturnsListIterator() {
+        assertTrue(of().listIterator(0) instanceof ListIterator);
+    }
+
+    @Test
+    public void testListIteratorWithIndex_ReturnsListIteratorWithCorrectPosition() {
+        Object object = new Object();
+
+        assertEquals(object, of(new Object(), object).listIterator(1).next());
+    }
+
+    @Test
+    public void testSubList_ReturnsImmutableList() {
+        Object object = new Object();
+
+        assertTrue(of(object, object).subList(0, 0) instanceof ImmutableList);
+    }
+
+    @Test
+    public void testSubList_ReturnImmutableListWithRightElements() {
+        Object object = new Object();
+
+        assertEquals(of(object), of(object, object).subList(0, 1));
+    }
+
+    @Test
+    public void testPlus_ReturnsImmutableList() {
+        assertTrue(of().plus(new Object()) instanceof ImmutableList);
+    }
+
+    @Test
+    public void testPlus_ReturnsNewInstanceOfImmutableList() {
+        ImmutableList<Object> list = of();
+
+        assertFalse(list == list.plus(new Object()));
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use assertSame(x, y) instead of assertTrue(x==y), or assertNotSame(x,y) vs assertFalse(x==y)](https://app.codacy.com/app/thomashan/coup-java/pullRequest?prid=4509131)